torch/_functorch/partitioners.py

```diff
@@ -296,13 +296,23 @@ def _has_tag_must_be_in_backward(node: fx.Node) -> bool:
 
 
 def _must_be_in_forward(node: fx.Node) -> bool:
-    return _has_tag_must_be_in_forward(node)
+    if _has_tag_must_be_in_forward(node):
+        return True
+    is_mutable = is_with_effects(node) or (
+        isinstance(node.target, torch._ops.OpOverload)
+        and node.target._schema.is_mutable
+    )
+    return not _has_tag_is_backward(node) and is_mutable
 
 
 def _must_be_in_backward(node: fx.Node) -> bool:
-    return _has_tag_must_be_in_backward(node) or (
-        _has_tag_is_backward(node) and is_with_effects(node)
+    if _has_tag_must_be_in_backward(node):
+        return True
+    is_mutable = is_with_effects(node) or (
+        isinstance(node.target, torch._ops.OpOverload)
+        and node.target._schema.is_mutable
     )
+    return _has_tag_is_backward(node) and is_mutable
```

Comment:
The extra torture tests for mutation that I added in `test_aotdispatch.py` caught some more problems: we use `_must_be_in_forward/backward` in the default partitioner [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/partitioners.py#L206) when deciding how to splice out the fw and bw graphs. If we have a side-effectful op, then a mutation must be in the fw if it was traced at fw time (and vis versa). previously, this check only accounted for WithEffects ops and not arbitrary mutable ops.

Reply:
This is very reasonable, but now I find myself a bit confused about how exactly the default partitioner works. Naively, I would have expected us to have just labeled all the forward nodes as forward and all the nodes traced during the backward() call as backwards, and then split this exactly. Since we don't do mutation in backwards, this would automatically cause all the mutations to go to forwards.